torch/optim/lbfgs.py

```diff
@@ -0,0 +1,243 @@
+        assert len(self.param_groups) == 1
+
+        group = self.param_groups[0]
+        lr = group['lr']
+        max_iter = group['max_iter']
+        max_eval = group['max_eval']
+        tolerance_grad = group['tolerance_grad']
+        tolerance_change = group['tolerance_change']
+        line_search_fn = group['line_search_fn']
+        history_size = group['history_size']
+
+        state = self.state['global_state']
+        state.setdefault('func_evals', 0)
+        state.setdefault('n_iter', 0)
+
+
+        # evaluate initial f(x) and df/dx
+        orig_loss = closure()
+        loss = orig_loss.data[0]
+        current_evals = 1
+        state['func_evals'] += 1
+
+        flat_grad = self._gather_flat_grad()
+        abs_grad_sum = flat_grad.abs().sum()
+
+        if abs_grad_sum <= tolerance_grad:
+            return loss
+
+        # variables cached in state (for tracing)
+        d = state.get('d')
+        t = state.get('t')
+        old_dirs = state.get('old_dirs')
+        old_stps = state.get('old_stps')
+        H_diag = state.get('H_diag')
+        prev_flat_grad = state.get('prev_flat_grad')
+        prev_loss = state.get('prev_loss')
+
+        n_iter = 0
+        # optimize for a max of max_iter iterations
+        while n_iter < max_iter:
+            # keep track of nb of iterations
+            n_iter += 1
+            state['n_iter'] += 1
+
+            ############################################################
+            # compute gradient descent direction
+            ############################################################
+            if state['n_iter'] == 1:
+                d = flat_grad.neg()
+                old_dirs = []
+                old_stps = []
+                H_diag = 1
+            else:
+                # do lbfgs update (update memory)
+                y = flat_grad.sub(prev_flat_grad)
+                s = d.mul(t)
+                ys = y.dot(s)  # y*s
+                if ys > 1e-10:
+                    # updating memory
+                    if len(old_dirs) == history_size:
+                        # shift history by one (limited-memory)
+                        old_dirs.pop(0)
+                        old_stps.pop(0)
+
+                    # store new direction/step
+                    old_dirs.append(s)
+                    old_stps.append(y)
+
+                    # update scale of initial Hessian approximation
+                    H_diag = ys / y.dot(y)  # (y*y)
+
+                # compute the approximate (L-BFGS) inverse Hessian
+                # multiplied by the gradient
+                num_old = len(old_dirs)
+
+                if 'ro' not in state:
+                    state['ro'] = [None] * history_size
+                    state['al'] = [None] * history_size
+                ro = state['ro']
+                al = state['al']
+
+                for i in range(num_old):
+                    ro[i] = 1 / old_stps[i].dot(old_dirs[i])
```

Comment:
can you make explicit all divisions on whether they are integer or floating.
here, do `1. /`